src/web/page/templates.rs

```diff
@@ -0,0 +1,191 @@
+use super::TEMPLATE_DATA;
+use crate::error::Result;
+use arc_swap::ArcSwap;
+use serde_json::Value;
+use std::collections::HashMap;
+use tera::{Result as TeraResult, Tera};
+
+/// Holds all data relevant to templating
+pub(crate) struct TemplateData {
+    /// The actual templates, stored in an `ArcSwap` so that they're hot-swappable
+    // TODO: Conditional compilation so it's not always wrapped, the `ArcSwap` is unneeded overhead for prod
+    pub templates: ArcSwap<Tera>,
+    /// The current global alert, serialized into a json value
+    global_alert: Value,
+    /// The version of docs.rs, serialized into a json value
+    docsrs_version: Value,
+    /// The current resource suffix of rustc, serialized into a json value
+    resource_suffix: Value,
+}
+
+impl TemplateData {
+    pub fn new() -> Result<Self> {
+        log::trace!("Loading templates");
+
+        let data = Self {
+            templates: ArcSwap::from_pointee(load_templates()?),
+            global_alert: serde_json::to_value(crate::GLOBAL_ALERT)?,
+            docsrs_version: Value::String(crate::BUILD_VERSION.to_owned()),
+            resource_suffix: Value::String(load_rustc_resource_suffix().unwrap_or_else(|err| {
+                log::error!("Failed to load rustc resource suffix: {:?}", err);
+                String::from("???")
+            })),
+        };
+
+        log::trace!("Finished loading templates");
+
+        Ok(data)
+    }
+
+    pub fn start_template_reloading() {
+        use std::{sync::Arc, thread, time::Duration};
+
+        thread::spawn(|| loop {
+            match load_templates() {
+                Ok(templates) => {
+                    log::info!("Reloaded templates");
+                    TEMPLATE_DATA.templates.swap(Arc::new(templates));
+                    thread::sleep(Duration::from_secs(10));
+                }
+
+                Err(err) => {
+                    log::info!("Error Loading Templates:\n{}", err);
+                    thread::sleep(Duration::from_secs(5));
+                }
+            }
+        });
+    }
+
+    /// Used to initialize a `TemplateData` instance in a `lazy_static`.
+    /// Loading tera takes a second, so it's important that this is done before any
+    /// requests start coming in
+    pub fn poke(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+// TODO: Is there a reason this isn't fatal? If the rustc version is incorrect (Or "???" as used by default), then
+//       all pages will be served *really* weird because they'll lack all CSS
```

Comment:
The only reason why this would fail to fetch the rustc version is if it's the first time you start docs.rs, before running `add-essential-files`. We can discuss (probably outside this PR) if that's actually the behavior we want.

Reply:
Do you want me to remove the comment?